service/history/workflow/mutable_state_impl.go

```diff
@@ -3335,6 +3351,7 @@ func (ms *MutableStateImpl) ApplyWorkflowExecutionFailedEvent(
 	ms.executionInfo.CompletionEventBatchId = firstEventID // Used when completion event needs to be loaded from database
 	ms.executionInfo.NewExecutionRunId = event.GetWorkflowExecutionFailedEventAttributes().GetNewExecutionRunId()
 	ms.executionInfo.CloseTime = event.GetEventTime()
+	ms.updateVisibilityVersionedTransition = true
```

Comment:
Updating this field everywhere is too error prone IMHO, we need to find a better way to set this. Can this be done just before closing the transaction?

Reply:
I created util func for updating search attributes and memo to set this field in the new PR. 
Those visibility update due to workflow state change, I moved the line for setting the field into SetWorkflowStateStatus.